### DISCUSSION TITLE
How does SEO works in nextjs?

### DISCUSSION THREAD

**Discussion by Danztee** [vercel/next.js#70066] [Help]
### Summary

I'm just thinking about how to do SEO in Next.js. Do we only need a sitemap and robots.txt file, along with using Google Search Console?

What do we need for better SEO in Next.js?

Also, HTML meta tags too?

Also, how does it compare to WordPress? What are your thoughts?

**Comment by SouthLink** [comment]
SEO is a very big topic for websites, next.js is just one part of it, the main aspects include:
1. Check your Robot.txt siteMap file
You should make sure that Google has access to the unique, high-quality pages you want to index. And block access to duplicate or private pages.
2. Do keyword research and set TDK correctly
TDK stands for title, description and keywords and plays a crucial role in optimizing web pages for search engines. If implemented effectively, TDK can significantly affect a website's ranking on the SERP.
3. Optimize your core network vitality
Core Web vitality metrics are page speed metrics that may affect Google rankings.
You can check your core network vitality in Google Search Console.
4. Mobile search engine optimization
Google uses mobile-first indexing, which means that the mobile version of your website is the "primary" version in Google's eyes.

You can also:
1. Page content is more easily crawled by Google bot through SSR technology
2. Configure a standard URL
3. Configure the ld+json correctly
...

Specific can refer to
https://developers.google.com/search/docs/fundamentals/seo-starter-guide

**Comment by condition123** [comment]
Hello,
Next.js is SEO-friendly because pages can be pre-rendered using SSR or SSG, allowing search engines to crawl content more efficiently. Technical factors like page speed, structured data, and clean site architecture still play a major role. I found some useful insights on local and technical SEO implementation at https://alfaranker.com/, especially for businesses aiming to improve organic visibility.
contact@alfaranker.com
+44 7438 598329
Abel st, Wyke Bar, Luxe Skincare, Bradford BD128BZ, United Kingdom

**Comment by Hassanmahmood4** [comment]
Sitemap + robots.txt + Search Console is a good start, but that's just the infrastructure layer. There's a lot more that goes into proper SEO in Next.js.

## The full checklist

### 1. Metadata (the most important part)

In App Router, use the built-in `metadata` export — no need for `next/head` or external libraries:

```js
// app/layout.js — global defaults
export const metadata = {
  title: {
    template: '%s | My Site',
    default: 'My Site',
  },
  description: 'A brief, compelling description under 160 characters.',
  openGraph: {
    type: 'website',
    locale: 'en_US',
    url: 'https://mysite.com',
    siteName: 'My Site',
    images: [{ url: '/og-image.png', width: 1200, height: 630 }],
  },
  twitter: {
    card: 'summary_large_image',
  },
};
```

```js
// app/blog/[slug]/page.js — per-page dynamic metadata
export async function generateMetadata({ params }) {
  const post = await getPost(params.slug);
  return {
    title: post.title,
    description: post.excerpt,
    openGraph: {
      title: post.title,
      description: post.excerpt,
      images: [{ url: post.coverImage }],
    },
  };
}
```

Every page needs a unique `title` and `description`. This is what actually shows up in Google results.

### 2. Sitemap and robots.txt

Next.js has native support — no packages needed:

```js
// app/sitemap.js
export default async function sitemap() {
  const posts = await getAllPosts();
  return [
    { url: 'https://mysite.com', lastModified: new Date(), priority: 1 },
    { url: 'https://mysite.com/about', lastModified: new Date(), priority: 0.8 },
    ...posts.map((post) => ({
      url: `https://mysite.com/blog/${post.slug}`,
      lastModified: post.updatedAt,
      priority: 0.6,
    })),
  ];
}
```

```js
// app/robots.js
export default function robots() {
  return {
    rules: { userAgent: '*', allow: '/' },
    sitemap: 'https://mysite.com/sitemap.xml',
  };
}
```

### 3. Structured data (JSON-LD)

This is what gets you rich results in Google (star ratings, FAQ dropdowns, breadcrumbs, etc.):

```js
// app/blog/[slug]/page.js
export default function BlogPost({ post }) {
  const jsonLd = {
    '@context': 'https://schema.org',
    '@type': 'Article',
    headline: post.title,
    datePublished: post.date,
    author: { '@type': 'Person', name: 'Your Name' },
    image: post.coverImage,
  };

  return (
    <>
      <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{ __html: JSON.stringify(jsonLd) }}
      />
      <article>{/* content */}</article>
    </>
  );
}
```

### 4. Technical SEO that people forget

- **Canonical URLs** — prevents duplicate content issues:
```js
  export const metadata = {
    alternates: { canonical: 'https://mysite.com/blog/my-post' },
  };
```
- **Semantic HTML** — use `<main>`, `<article>`, `<section>`, `<h1>` (one per page), `<h2>`, `<h3>` in order. Google parses these.
- **Image alt text** — every `<Image>` needs a descriptive `alt` prop. Not optional for SEO.
- **Loading performance** — use `next/image` (auto lazy loading, WebP conversion), `next/font` (no layout shift from fonts), and keep your bundle lean. Core Web Vitals (LCP, CLS, INP) are a direct ranking factor.
- **Internal linking** — link between your own pages naturally. This is how Google discovers and ranks your content hierarchy.
- **Clean URLs** — `/blog/my-post` not `/blog?id=123`.

### 5. Rendering strategy matters

| Strategy | SEO | When to use |
|---|---|---|
| SSG (`generateStaticParams`) | Best — HTML ready at crawl time | Blog posts, docs, marketing pages |
| SSR (dynamic server components) | Great — HTML generated on request | Frequently changing content |
| CSR (client components with `useEffect`) | Bad — Google may not index it | Dashboards, logged-in pages (SEO irrelevant) |

For anything you want indexed, make sure it renders on the server. If your content loads inside a `useEffect` or after a client-side fetch, Google might see a blank page.

## Next.js vs WordPress for SEO

Honestly, WordPress still has an edge in **convenience** — install Yoast/RankMath and you get a guided checklist, auto-sitemaps, schema markup, social previews, and readability scoring out of the box. A non-technical person can manage SEO without touching code.

But Next.js has the edge in **performance and control**:
- Faster Core Web Vitals out of the box (WordPress sites are typically heavier with plugins/themes)
- Full control over rendering strategy per page
- No plugin bloat slowing things down
- You can implement everything WordPress plugins do — you just write it yourself

**Bottom line:** WordPress is easier for content-heavy sites where non-devs manage SEO. Next.js is better when you want top performance scores and full control, and you're comfortable implementing SEO features in code. The actual ranking factors (content quality, backlinks, metadata, page speed) are the same regardless of framework.

## TL;DR

Sitemap + robots.txt + Search Console is just step one. The real SEO work is: unique metadata per page, structured data for rich results, semantic HTML, fast Core Web Vitals, server-side rendering for indexable pages, and canonical URLs. Next.js gives you all the tools natively — no external packages needed.

**Comment by KoichiNakamura038** [comment]
SEO in Next.js is similar to SEO in any modern website, but Next.js provides several built-in advantages.

For better SEO, I recommend focusing on:

1. Metadata
   - Page title
   - Meta description
   - Canonical URL
   - Open Graph tags
   - Twitter Card tags

2. Sitemap
   - Generate a sitemap.xml so search engines can easily discover your pages.

3. robots.txt
   - Allow search engines to crawl public content.
   - Block pages that should not be indexed.

4. SSR or SSG
   - One of Next.js's biggest SEO advantages is server-side rendering (SSR) and static site generation (SSG), which make content easier for search engines to crawl and index.

5. Structured Data
   - Add JSON-LD schema markup when relevant (articles, products, FAQs, etc.).

6. Performance
   - Optimize Core Web Vitals.
   - Use next/image, lazy loading, and code splitting to improve page speed.

7. Google Search Console
   - Submit your sitemap.
   - Monitor indexing, search performance, and crawl issues.

Compared to WordPress:

- WordPress offers many SEO features through plugins like Yoast SEO or Rank Math.
- Next.js requires more manual setup but provides greater flexibility, performance, and control.
- WordPress is often easier for content-heavy blogs.
- Next.js is often preferred for modern web applications, SaaS platforms, and highly customized websites.

In short, sitemap.xml, robots.txt, and Google Search Console are important, but metadata, rendering strategy (SSR/SSG), structured data, and performance optimization are also key factors for strong SEO in Next.js.

**Comment by Saif0933** [comment]
SEO in Next.js is not just about adding a sitemap, robots.txt, and submitting your site to Google Search Console. Those are important, but they're only part of the overall SEO strategy.

Here are the areas I would focus on:

1. Metadata (Very Important)

Make sure every page has unique:

Title
Meta description
Open Graph tags
Twitter card tags
Canonical URL

In modern Next.js, you can use the Metadata API:

export const metadata = {
  title: "My Page",
  description: "Description of my page",
};
2. Server-Side Rendering and Static Generation

One advantage of Next.js is that pages can be pre-rendered.

Use:

Static Generation (SSG) whenever possible
Server-Side Rendering (SSR) when content changes frequently

This helps search engines discover content faster compared to a client-side rendered application.

3. Sitemap and robots.txt

These help search engines understand:

Which pages should be indexed
Which pages should be ignored

For larger sites, keeping the sitemap updated automatically is very helpful.

4. Core Web Vitals

Google cares about user experience.

Focus on:

Fast page load times
Optimized images (next/image)
Reduced JavaScript bundle size
Good Largest Contentful Paint (LCP)
Low Cumulative Layout Shift (CLS)

You can monitor these in:

Google Search Console
PageSpeed Insights
Lighthouse
5. Structured Data (JSON-LD)

Adding structured data helps Google better understand your content.

Examples:

Articles
Products
FAQs
Reviews
Organizations

This can improve your chances of getting rich results in search.

6. Content Quality

No framework can compensate for poor content.

SEO depends heavily on:

Solving user problems
Targeting relevant keywords
Creating useful and original content
Internal linking between related pages

Content quality usually has a bigger impact than technical SEO alone.

7. Mobile Optimization

Google uses mobile-first indexing.

Make sure:

Responsive design works well
Fonts are readable
Navigation is easy on mobile devices
Pages load quickly on slower networks
8. Google Search Console

After deployment:

Submit your sitemap
Monitor indexing issues
Check Core Web Vitals
Track search performance
Next.js vs WordPress

I don't think one is universally "better" for SEO.

WordPress provides many SEO features through plugins like Yoast SEO or Rank Math, making it easier for non-developers.
Next.js gives developers much more control over rendering, performance, metadata, and architecture.

If implemented correctly, both can rank extremely well in search results. The deciding factors are usually content quality, site performance, user experience, and overall SEO strategy rather than the framework itself.

In my experience, Next.js gives more flexibility and performance optimization options, while WordPress offers a faster SEO setup through its plugin ecosystem

**Comment by anmoldwivedi89** [comment]
# SEO in Next.js

## 1. Meta Tags (Head Component)
- Title, Description, OG tags

## 2. Sitemap.xml
- Automatic sitemap generation

## 3. Robots.txt
- Search engine crawling

## 4. Dynamic Meta Tags
- Use next/head या next/seo

## 5. Image Optimization
- next/image for better SEO

## 6. Performance
- Faster loading = Better SEO
- import Head from 'next/head'

export default function Page() {
  return (
    <>
      <Head>
        <title>SEO Title</title>
        <meta name="description" content="..." />
        <meta property="og:title" content="..." />
      </Head>
      <h1>Content</h1>
    </>
  )
}

**Comment by kuldeeprajput-dev** [comment]
SEO in Next.js is more than just `sitemap.xml`, `robots.txt`, and Google Search Console.

For good SEO, you should focus on:

1. Use SSR, SSG, or ISR so pages are crawlable.
2. Add proper metadata like title, description, canonical URL, and Open Graph tags.
3. Use `sitemap.xml` and `robots.txt`.
4. Structure content properly with headings, clean URLs, internal links, and image alt text.
5. Optimize performance, images, fonts, and Core Web Vitals.
6. Add JSON-LD structured data when needed.
7. Use Google Search Console to monitor indexing and search issues.

Compared to WordPress, WordPress is easier for basic SEO because plugins handle many things automatically. Next.js gives more control and performance, but you need to set up SEO properly yourself.
SEO in Next.js is more than just `sitemap.xml`, `robots.txt`, and Google Search Console.

For good SEO, you should focus on:

1. Use SSR, SSG, or ISR so pages are crawlable.
2. Add proper metadata like title, description, canonical URL, and Open Graph tags.
3. Use `sitemap.xml` and `robots.txt`.
4. Structure content properly with headings, clean URLs, internal links, and image alt text.
5. Optimize performance, images, fonts, and Core Web Vitals.
6. Add JSON-LD structured data when needed.
7. Use Google Search Console to monitor indexing and search issues.

Compared to WordPress, WordPress is easier for basic SEO because plugins handle many things automatically. Next.js gives more control and performance, but you need to set up SEO properly yourself.
